// sql-ts/dist/node/column.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
class ColumnNode extends _1.Node {
    constructor(config) {
        super('COLUMN');
        this.name = config.name;
        this.property = config.property || config.name;
        this.alias = config.alias;
        this.star = config.star;
        this.isConstant = config.isConstant;
        this.constantValue = config.constantValue;
        this.asArray = config.asArray;
        this.aggregator = config.aggregator;
        this.table = config.table;
        this.value = config.getValue();
        this.dataType = config.dataType;
        this.isDistinct = config.isDistinct;
        this.primaryKey = config.primaryKey;
        this.notNull = config.notNull;
        this.defaultValue = config.defaultValue;
        this.references = config.references;
        // If subfieldContainer is present, this is a subfield and subfieldContainer
        // is the parent Column
        this.subfieldContainer = config.subfieldContainer;
        this.subfields = config.subfields;
        this.autoGenerated = !!config.autoGenerated;
        this.unique = !!config.unique;
    }
    distinct() {
        this.isDistinct = true;
        return this;
    }
    as(alias) {
        this.alias = alias;
        return this;
    }
}
exports.ColumnNode = ColumnNode;
