// sql-ts/dist/node/node.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const _1 = require(".");
const dialect_1 = require("../dialect");
const nodeable_1 = require("../nodeable");
class Node {
    constructor(type) {
        this.type = type;
        this.nodes = [];
    }
    toNode() {
        return this;
    }
    add(node) {
        assert(node, 'Error while trying to add a non-existant node to a query');
        let newNode;
        if (typeof node === 'string') {
            newNode = new _1.TextNode(node);
        }
        else if (nodeable_1.instanceofINodeable(node)) {
            newNode = node.toNode();
        }
        else {
            newNode = node;
        }
        this.nodes.push(newNode);
        return this;
    }
    unshift(node) {
        assert(node, 'Error while trying to add a non-existant node to a query');
        this.nodes.unshift(typeof node === 'string' ? new _1.TextNode(node) : node.toNode());
        return this;
    }
    toQuery(dialect) {
        const DialectClass = determineDialect(this, dialect);
        return initializeDialect(DialectClass, this).getQuery(this);
    }
    toNamedQuery(name, dialect) {
        if (!name || typeof name !== 'string' || name === '') {
            throw new Error('A query name has to be a non-empty String.');
        }
        const query = this.toQuery(dialect);
        return Object.assign({}, query, { name });
    }
    toString(dialect) {
        const DialectClass = determineDialect(this, dialect);
        return initializeDialect(DialectClass, this).getString(this);
    }
    addAll(nodes) {
        for (let i = 0, len = nodes.length; i < len; i++) {
            this.add(nodes[i]);
        }
        return this;
    }
}
exports.Node = Node;
// Before the change that introduced parallel dialects, every node could be turned
// into a query. The parallel dialects change made it impossible to change some nodes
// into a query because not all nodes are constructed with the sql instance.
const determineDialect = (query, dialect) => {
    const sql = query.sql || (query.table && query.table.sql);
    let DialectClass;
    if (dialect) {
        // dialect is specified
        DialectClass = dialect_1.getDialect(dialect);
    }
    else if (sql && sql.dialect) {
        // dialect is not specified, use the dialect from the sql instance
        DialectClass = sql.dialect;
    }
    else {
        // dialect is not specified, use the default dialect
        DialectClass = dialect_1.getDialect(dialect_1.DEFAULT_DIALECT);
    }
    return DialectClass;
};
const initializeDialect = (DialectClass, query) => {
    const sql = query.sql || (query.table && query.table.sql);
    const config = sql ? sql.config : {};
    return new DialectClass(config);
};
