// sql-ts/dist/dialect/oracle.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const postgres_1 = require("./postgres");
class Oracle extends postgres_1.Postgres {
    constructor(config) {
        super(config);
        this.myClass = Oracle;
        this.aliasText = ' ';
    }
    _getParameterValue(value, quoteChar) {
        return Buffer.isBuffer(value)
            ? "utl_raw.cast_to_varchar2(hextoraw('" + value.toString('hex') + "'))"
            : super._getParameterValue(value);
    }
    _getParameterPlaceholder(index, value) {
        return `:${index}`;
    }
    visitReplace(replaceNode) {
        throw new Error('Oracle does not support REPLACE.');
    }
    visitAlias(aliasNode) {
        const result = [this.visit(aliasNode.value) + ' ' + this.quote(aliasNode.alias)];
        return result;
    }
    visitAlter(alterNode) {
        const errMsg = 'ALTER TABLE cannot be used to perform multiple different operations in the same statement.';
        // Implement our own add column:
        //   PostgreSQL: ALTER TABLE "name" ADD COLUMN "col1", ADD COLUMN "col2"
        //   Oracle:  ALTER TABLE "name" ADD ("col1", "col2")
        const addColumn = () => {
            this.visitingAlter = true;
            const table = this.queryNode.table;
            this.visitingAddColumn = true;
            let result = 'ALTER TABLE ' + this.visit(table.toNode()) + ' ADD (' + this.visit(alterNode.nodes[0].nodes[0]);
            for (let i = 1, len = alterNode.nodes.length; i < len; i++) {
                const node = alterNode.nodes[i];
                assert(node.type === 'ADD COLUMN', errMsg);
                result += ', ' + this.visit(node.nodes[0]);
            }
            result += ')';
            this.visitingAddColumn = false;
            this.visitingAlter = false;
            return [result];
        };
        // Implement our own drop column:
        //   PostgreSQL: ALTER TABLE "name" DROP COLUMN "col1", DROP COLUMN "col2"
        //   Oracle:  ALTER TABLE "name" DROP ("col1", "col2")
        const dropColumn = () => {
            this.visitingAlter = true;
            const table = this.queryNode.table;
            const result = ['ALTER TABLE', ...this.visit(table.toNode())];
            let columns = 'DROP (' + this.visit(alterNode.nodes[0].nodes[0]);
            for (let i = 1, len = alterNode.nodes.length; i < len; i++) {
                const node = alterNode.nodes[i];
                assert(node.type === 'DROP COLUMN', errMsg);
                columns += ', ' + this.visit(node.nodes[0]);
            }
            columns += ')';
            result.push(columns);
            this.visitingAlter = false;
            return result;
        };
        if (isAlterAddColumn(alterNode)) {
            return addColumn();
        }
        if (isAlterDropColumn(alterNode)) {
            return dropColumn();
        }
        return super.visitAlter(alterNode);
    }
    visitTable(tableNode) {
        const table = tableNode.table;
        let txt = '';
        if (table.getSchema()) {
            txt = this.quote(table.getSchema());
            txt += '.';
        }
        txt += this.quote(table.getName());
        if (table.alias) {
            txt += ' ' + this.quote(table.alias);
        }
        return [txt];
    }
    visitCascade(cascadeNode) {
        return ['CASCADE CONSTRAINTS'];
    }
    visitRestrict(restrictNode) {
        throw new Error('Oracle do not support RESTRICT in DROP TABLE');
    }
    visitDrop(dropNode) {
        if (!isDropIfExists(dropNode)) {
            return super.visitDrop(dropNode);
        }
        // Implement our own drop if exists:
        //   PostgreSQL: DROP TABLE IF EXISTS "group"
        //   Oracle:
        //     BEGIN
        //          EXECUTE IMMEDIATE 'DROP TABLE POST';
        //     EXCEPTION
        //          WHEN OTHERS THEN
        //                 IF SQLCODE != -942 THEN
        //                      RAISE;
        //                 END IF;
        //     END;
        const table = this.queryNode.table;
        const tableResult = this.visit(table.toNode());
        const dropResult = ['DROP TABLE'];
        dropResult.push(...tableResult);
        return [
            "BEGIN EXECUTE IMMEDIATE '" + dropResult.join(' ') + "'; EXCEPTION WHEN OTHERS THEN IF SQLCODE != -942 THEN RAISE; END IF; END;"
        ];
    }
    visitCreate(createNode) {
        const isNotExists = isCreateIfNotExists(createNode);
        // const isTemporary=isCreateTemporary(create)
        let createText = super.visitCreate(createNode);
        if (isNotExists) {
            // Implement our own create if not exists:
            //   PostgreSQL: CREATE TABLE IF NOT EXISTS "group" ("id" constchar(100))
            //   Oracle:
            //     BEGIN
            //          EXECUTE IMMEDIATE 'CREATE TABLE ...';
            //     EXCEPTION
            //          WHEN OTHERS THEN
            //                 IF SQLCODE != -955 THEN
            //                      RAISE;
            //                 END IF;
            //     END;
            createText = [
                "BEGIN EXECUTE IMMEDIATE '" +
                    createText.join(' ').replace(' IF NOT EXISTS', '') +
                    "'; EXCEPTION WHEN OTHERS THEN IF SQLCODE != -955 THEN RAISE; END IF; END;"
            ];
        }
        return createText;
    }
    visitBinary(binaryNode) {
        if (binaryNode.operator === '@@') {
            let text = '(INSTR (' + this.visit(binaryNode.left) + ', ';
            text += this.visit(binaryNode.right);
            text += ') > 0)';
            return [text];
        }
        if (!isRightSideArray(binaryNode)) {
            return super.visitBinary(binaryNode);
        }
        if (binaryNode.operator === 'IN' || binaryNode.operator === 'NOT IN') {
            return super.visitBinary(binaryNode);
        }
        throw new Error('Oracle does not support arrays in this type of expression.');
    }
    visitModifier(modifierNode) {
        const ret = super.visitModifier(modifierNode);
        if (ret.indexOf('OFFSET') >= 0) {
            ret.push('ROWS');
        }
        if (ret.indexOf('LIMIT') >= 0) {
            ret[0] = 'FETCH NEXT';
            ret.push('ROWS ONLY');
        }
        return ret;
    }
    visitQueryHelper(actions, targets, filters) {
        const output = super.visitQueryHelper(actions, targets, filters);
        // In Oracle, OFFSET must come before FETCH NEXT (limit)
        // Change positions, if both are present and not done already
        const offset = output.indexOf('OFFSET');
        const limit = output.indexOf('FETCH NEXT');
        if (offset !== -1 && limit !== -1 && offset > limit) {
            const temp = [output[offset], output[offset + 1], output[offset + 2]];
            output[offset] = output[limit];
            output[offset + 1] = output[limit + 1];
            output[offset + 2] = output[limit + 2];
            output[limit] = temp[0];
            output[limit + 1] = temp[1];
            output[limit + 2] = temp[2];
        }
        return this.output;
    }
    visitColumn(columnNode) {
        const table = columnNode.table;
        const inSelectClause = !this.selectOrDeleteEndIndex;
        const arrayAgg = () => {
            throw new Error('Oracle does not support array_agg.');
        };
        const countStar = () => {
            // Implement our own since count(table.*) is invalid in Oracle
            let result = 'COUNT(*)';
            if (inSelectClause && columnNode.alias) {
                result += this.aliasText + this.quote(columnNode.alias);
            }
            return [result];
        };
        if (isCountStarExpression(columnNode)) {
            return countStar();
        }
        if (inSelectClause && table && !table.alias && columnNode.asArray) {
            return arrayAgg();
        }
        return super.visitColumn(columnNode);
    }
    visitReturning(returningNode) {
        // TODO: need to add some code to the INSERT clause to support this since its the equivalent of the OUTPUT clause
        // in MS SQL which appears before the values, not at the end of the statement.
        throw new Error('Returning clause is not yet supported for Oracle.');
    }
    visitIndexes(indexesNode) {
        const tableName = this.queryNode.table.getName();
        const schemaName = this.queryNode.table.getSchema();
        let indexes = "SELECT * FROM USER_INDEXES WHERE TABLE_NAME = '" + tableName + "'";
        if (schemaName) {
            indexes += " AND TABLE_OWNER = '" + schemaName + "'";
        }
        return [indexes];
    }
    visitDropIndex(dropIndexNode) {
        const result = ['DROP INDEX'];
        const schemaName = dropIndexNode.table.getSchema();
        if (schemaName) {
            result.push(this.quote(schemaName) + '.');
        }
        result.push(this.quote(dropIndexNode.options.indexName));
        return result;
    }
    // Need to implement a special version of CASE since Oracle doesn't support
    //   CASE WHEN true THEN xxx END
    //   the "true" has to be a boolean expression like 1=1
    visitCase(caseNode) {
        const whenValue = (node) => {
            if (node.type !== 'PARAMETER') {
                return this.visit(node);
            }
            // dealing with a true/false value
            const val = node.value();
            if (val === true) {
                return '1=1';
            }
            else {
                return '0=1';
            }
        };
        assert(caseNode.whenList.length === caseNode.thenList.length);
        let text = '(CASE';
        this.visitingCase = true;
        for (let i = 0; i < caseNode.whenList.length; i++) {
            text += ` WHEN ${whenValue(caseNode.whenList[i])} THEN ${this.visit(caseNode.thenList[i])}`;
        }
        if (caseNode.else != null) {
            text += ` ELSE ${this.visit(caseNode.else)}`;
        }
        this.visitingCase = false;
        text += ' END)';
        return [text];
    }
    visitOnConflict(onConflictNode) {
        throw new Error('Oracle does not allow onConflict clause.');
    }
}
exports.Oracle = Oracle;
function isCreateIfNotExists(createNode) {
    if (createNode.nodes.length === 0) {
        return false;
    }
    if (createNode.nodes[0].type !== 'IF NOT EXISTS') {
        return false;
    }
    return true;
}
function isCreateTemporary(createNode) {
    return createNode.options.isTemporary;
}
function isDropIfExists(dropNode) {
    if (dropNode.nodes.length === 0) {
        return false;
    }
    if (dropNode.nodes[0].type !== 'IF EXISTS') {
        return false;
    }
    return true;
}
// SQL Server does not support array expressions except in the IN clause.
function isRightSideArray(binaryNode) {
    return Array.isArray(binaryNode.right);
}
function isCountStarExpression(columnNode) {
    if (!columnNode.aggregator) {
        return false;
    }
    if (columnNode.aggregator.toLowerCase() !== 'count') {
        return false;
    }
    if (!columnNode.star) {
        return false;
    }
    return true;
}
function isAlterAddColumn(alterNode) {
    if (alterNode.nodes.length === 0) {
        return false;
    }
    if (alterNode.nodes[0].type !== 'ADD COLUMN') {
        return false;
    }
    return true;
}
function isAlterDropColumn(alterNode) {
    if (alterNode.nodes.length === 0) {
        return false;
    }
    if (alterNode.nodes[0].type !== 'DROP COLUMN') {
        return false;
    }
    return true;
}
